### PR TITLE
T23173: Drop boost dependency, port to Meson

### PR DESCRIPTION
**Pull request by smspillaz**
This is part of a task to make this code more maintainable for Endless (the main user of this project).

For various reasons having a dependency on boost and cmake makes distro maintenance a bit of a nightmare, so this PR gets rid of both of those, replacing boost template code with handwritten templates and cmake with meson.

The main part of this work was to drop the dependency on `boost::geometry`. This was trickier than it looked, as we were actually using boost::geometry to support arithmetic between array-of-structures  (`PointModel`) and structure-of-arrays (`PointView`) types. So, in order to preserve the structure-of-arrays optimizations, this PR re-implements some of the `boost::geometry` type traits and operations in `wobbly/geometry_traits`. The main concept there is that both `PointView` and `PointModel` implement the `wobbly::geometry::dimension:Dimension` and `wobbly::geometry::dimension:DimensionAccess` type trait, meaning template substitution will allow generating code for those types when passed to anything in `wobbly::geometry::dimension`.

Finally, we drop `cmake` and all of the polysquare-specific (now unmaintained by yours truly) code and switch to meson. This sadly means that we lose support for some tooling, but meson is rapidly [gaining support for that anyway](http://mesonbuild.com/Release-notes-for-0-42-0.html#page-description).

https://phabricator.endlessm.com/T23173

### CODE REVIEW

File: src/wobbly.cpp [hunk 1]
```diff
@@ -64,7 +64,7 @@
 #include <type_traits>                  // for move, etc
 #include <vector>                       // for vector
 
-#include <boost/optional/optional.hpp>
+#include <experimental/optional>        // for optional
```
Comment (smspillaz) [inline src/wobbly.cpp]: This header can even be dropped.

File: include/wobbly/geometry.h [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+                public:
+
+                    CopyablePV (CopyablePV const &copy) = default;
+                    CopyablePV & operator= (CopyablePV const &copy) = default;
+            };
+
+            struct NoncopyablePV
+            {
+                public:
+
```
Comment (joaquimrocha) [inline include/wobbly/geometry.h]: This and the above `public` have a newline below but not the `PointViewCopyabilityBase` class. I prefer it without the newline, but I am not sure you just copied it from somewhere and wanted to keep it like this for some reason.

File: include/wobbly/geometry.h [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                PointView (PointView <NumericType> const &p) :
+                    array (p.array),
+                    offset (p.offset)
+                {
+                }
+
+                friend void swap (PointView <NumericType> &first,
+                                  PointView <NumericType> &second)
+                {
+                    using std::swap;
```
Comment (joaquimrocha) [inline include/wobbly/geometry.h]: Nitpick: I'd prefer to just use `std::swap` in the calls below. Makes it clearer.

File: include/wobbly/geometry.h [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+                template <size_t Offset>
+                typename std::enable_if <!(IC::value) && (Offset + 1), void>::type
+                set (NumericType value)
+                {
+                    static_assert (Offset < 2, "Offset must be < 2");
+                    array[offset + Offset] = value;
+                }
+
+            private:
+
```
Comment (joaquimrocha) [inline include/wobbly/geometry.h]: Nitpick: The newline again here.

File: include/wobbly/geometry.h [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+            PointModel (PointModel const &p) noexcept :
+                x (p.x),
+                y (p.y)
+            {
+            }
+
+            void swap (PointModel &a, PointModel &b) noexcept
+            {
+                using std::swap;
+
```
Comment (joaquimrocha) [inline include/wobbly/geometry.h]: Again, I'd prefer to use the namespace here. The use of the newline also differs from the prevuous swap where there was none.

File: include/wobbly/geometry_traits.h [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+                        set <0> (dst, v (get <0> (dst), src));
+                    }
+                };
+
+                template <typename T, typename U, typename A, typename Visitor, size_t D>
+                struct ReduceCoordinatePairsRecursive
+                {
+                    static A apply (T const &l,
+                                    U const &r,
+                                    A       &&acc,
```
Comment (joaquimrocha) [inline include/wobbly/geometry_traits.h]: Nitpick: The alignment here differs from the rest of the file.

File: include/wobbly/geometry.h (hunk 5 above)
Comment (smspillaz) [reply include/wobbly/geometry.h]: Sure, make sense

File: include/wobbly/geometry.h (hunk 3 above)
Comment (smspillaz) [reply include/wobbly/geometry.h]: Makes sense

File: include/wobbly/geometry.h (hunk 4 above)
Comment (smspillaz) [reply include/wobbly/geometry.h]: The newline is part of the style of the file, I think.

File: include/wobbly/geometry.h (hunk 2 above)
Comment (smspillaz) [reply include/wobbly/geometry.h]: Ah, I should adjust `PointViewCopyabilityBase`. I am pretty sure there is meant to be a newline.

**Comment by smspillaz** [comment]
> Maybe a more descriptive commit title and message in the wobbly: Use noexcept(false) commit? There are other points in the file where noexcept is already used.

> There's a typo in the commit's body fromt the type traits. in include: Add hand-written geometry type traits and types.

Both comments make sense, I'll fix up the commit messages

File: include/wobbly/geometry_traits.h (hunk 6 above)
Comment (smspillaz) [reply include/wobbly/geometry_traits.h]: Does it? Not sure what you mean here.

**Pull request merged**